atlasdb-config/src/test/java/com/palantir/atlasdb/http/SerializableErrorDecoderTest.java

```diff
@@ -0,0 +1,131 @@
+
+    private static final String EMPTY_REASON = "";
+    private static final byte[] EMPTY_BODY = new byte[0];
+    private static final String SOME_METHOD_KEY = "METHOD";
+    private static final String EXCEPTION_MESSAGE = "foo";
+    private static final String LOCK_ID = "lock";
+
+    private static final String KNOWN_ERROR_NAME = IllegalArgumentException.class.getCanonicalName();
+    private static final String UNKNOWN_ERROR_NAME = "com.palantir.atlasdb.errors.one.two.three.four";
+    private static final String NOT_ERROR_NAME = "java.lang.Integer";
+
+    private final SerializableErrorDecoder decoder = new SerializableErrorDecoder();
+
+    @Test
+    public void canDecodeExceptionOfKnownType() throws JsonProcessingException {
+        SerializableError serializableError = SerializableError.of(EXCEPTION_MESSAGE, KNOWN_ERROR_NAME);
+        Response response = createResponseForEntity(serializableError);
+
+        Exception exception = decoder.decode(SOME_METHOD_KEY, response);
+        assertThat(exception).isInstanceOf(AtlasDbRemoteException.class)
+                .satisfies(remoteException -> assertHasErrorName(remoteException, KNOWN_ERROR_NAME));
+    }
+
+    @Test
+    public void canDecodeExceptionOfUnknownType() throws JsonProcessingException {
+        SerializableError serializableError = SerializableError.of(EXCEPTION_MESSAGE, UNKNOWN_ERROR_NAME);
+        Response response = createResponseForEntity(serializableError);
+
+        Exception exception = decoder.decode(SOME_METHOD_KEY, response);
+        assertThat(exception).isInstanceOf(AtlasDbRemoteException.class)
+                .satisfies(remoteException -> assertHasErrorName(remoteException, UNKNOWN_ERROR_NAME));
+    }
+
+    @Test
+    public void canDecodeExceptionBelievedToNotBeAnExceptionType() throws JsonProcessingException {
+        SerializableError serializableError = SerializableError.of(EXCEPTION_MESSAGE, NOT_ERROR_NAME);
+        Response response = createResponseForEntity(serializableError);
+
+        Exception exception = decoder.decode(SOME_METHOD_KEY, response);
+        assertThat(exception).isInstanceOf(AtlasDbRemoteException.class)
+                .satisfies(remoteException -> assertHasErrorName(remoteException, NOT_ERROR_NAME));
+    }
+
+    @Test
+    public void resilientToEmptyBody() {
+        Response response = createResponse(HttpStatus.SC_SERVICE_UNAVAILABLE, EMPTY_BODY);
+
+        assertCanDecodeRuntimeException(response);
+    }
+
+    @Test
+    public void resilientToMalformedBody() {
+        byte[] oneByteArray = {42};
+        Response response = createResponse(HttpStatus.SC_SERVICE_UNAVAILABLE, oneByteArray);
+
+        assertCanDecodeRuntimeException(response);
+    }
+
+    @Test
+    public void resilientToValidJsonBodyThatIsNotASerializableError() throws JsonProcessingException {
+        LockRequest lockRequest =
+                LockRequest.builder(ImmutableSortedMap.of(StringLockDescriptor.of(LOCK_ID), LockMode.WRITE))
```

Comment:
(Let's discuss this offline - the idea is actually that the AtlasDbErrorDecoder is the 'main' decoder that calls this first!)